File: tests/fixtures/Sample.java
/**
 * Java test fixture for Tree-sitter parsing tests.
 * 
 * This file contains various Java language constructs to test comprehensive parsing.
 */

package com.example.fixtures;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;
import java.io.*;
import java.nio.file.*;
import javax.annotation.processing.*;
import org.springframework.stereotype.*;
import org.springframework.beans.factory.annotation.*;

/**
 * Configuration class with builder pattern.
 */
@Component
@ConfigurationProperties(prefix = "processor")
public class ProcessorConfiguration {
    private String name = "default";
    private boolean debug = false;
    private int timeout = 5000;
    private List<String> allowedTypes = new ArrayList<>();
    
    // Constructor
    public ProcessorConfiguration() {}
    
    public ProcessorConfiguration(String name, boolean debug) {
        this.name = name;
        this.debug = debug;
    }
    
    // Getters and setters
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public boolean isDebug() { return debug; }
    public void setDebug(boolean debug) { this.debug = debug; }
    
    public int getTimeout() { return timeout; }
    public void setTimeout(int timeout) { this.timeout = timeout; }
    
    public List<String> getAllowedTypes() { return new ArrayList<>(allowedTypes); }
    public void setAllowedTypes(List<String> allowedTypes) { 
        this.allowedTypes = new ArrayList<>(allowedTypes); 
    }
    
    /**
     * Builder pattern implementation.
     */
    public static class Builder {
        private String name = "default";
        private boolean debug = false;
        private int timeout = 5000;
        private List<String> allowedTypes = new ArrayList<>();
        
        public Builder name(String name) {
            this.name = name;
            return this;
        }
        
        public Builder debug(boolean debug) {
            this.debug = debug;
            return this;
        }
        
        public Builder timeout(int timeout) {
            this.timeout = timeout;
            return this;
        }
        
        public Builder allowedType(String type) {
            this.allowedTypes.add(type);
            return this;
        }
        
        public ProcessorConfiguration build() {
            ProcessorConfiguration config = new ProcessorConfiguration(name, debug);
            config.setTimeout(timeout);
            config.setAllowedTypes(allowedTypes);
            return config;
        }
    }
    
    public static Builder builder() {
        return new Builder();
    }
}

/**
 * Abstract base class for data processors.
 */
public abstract class BaseDataProcessor<T> {
    protected final ProcessorConfiguration config;
    protected final List<T> results;
    protected int errorCount;
    
    public BaseDataProcessor(ProcessorConfiguration config) {
        this.config = Objects.requireNonNull(config, "Configuration cannot be null");
        this.results = Collections.synchronizedList(new ArrayList<>());
        this.errorCount = 0;
    }
    
    /**
     * Abstract method for processing individual items.
     * @param item The item to process
     * @return The processed result
     * @throws ProcessingException if processing fails
     */
    public abstract T processItem(String item) throws ProcessingException;
    
    /**
     * Process a collection of items.
     * @param items Items to process
     * @return List of processed results
     */
    public List<T> process(Collection<String> items) {
        Objects.requireNonNull(items, "Items cannot be null");
        
        List<T> processed = new ArrayList<>();
        
        for (String item : items) {
            try {
                T result = processItem(item);
                processed.add(result);
                results.add(result);
                
                if (config.isDebug()) {
                    System.out.println("Processed: " + item + " -> " + result);
                }
            } catch (ProcessingException e) {
                errorCount++;
                handleError(e, item);
            }
        }
        
        return processed;
    }
    
    /**
     * Process items asynchronously using CompletableFuture.
     * @param items Items to process
     * @return CompletableFuture with list of results
     */
    public CompletableFuture<List<T>> processAsync(Collection<String> items) {
        return CompletableFuture.supplyAsync(() -> {
            return items.parallelStream()
                    .map(item -> {
                        try {
                            return processItem(item);
                        } catch (ProcessingException e) {
                            handleError(e, item);
                            return null;
                        }
                    })
                    .filter(Objects::nonNull)
                    .collect(Collectors.toList());
        });
    }
    
    /**
     * Handle processing errors.
     * @param error The error that occurred
     * @param item The item that caused the error
     */
    protected void handleError(ProcessingException error, String item) {
        if (config.isDebug()) {
            System.err.println("Error processing '" + item + "': " + error.getMessage());
        }
    }
    
    /**
     * Get processing statistics.
     * @return Map of statistics
     */
    public Map<String, Object> getStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalProcessed", results.size());
        stats.put("errorCount", errorCount);
        stats.put("configName", config.getName());
        return stats;
    }
    
    // Protected methods for subclasses
    protected boolean validateItem(String item) {
        return item != null && !item.trim().isEmpty();
    }
    
    protected void logDebug(String message) {
        if (config.isDebug()) {
            System.out.println("[DEBUG] " + message);
        }
    }
}

/**
 * Concrete implementation of BaseDataProcessor for string processing.
 */
@Service
public class StringDataProcessor extends BaseDataProcessor<String> {
    
    private final ExecutorService executorService;
    
    @Autowired
    public StringDataProcessor(ProcessorConfiguration config) {
        super(config);
        this.executorService = Executors.newFixedThreadPool(4);
    }
    
    // Alternative constructor for testing
    public StringDataProcessor(ProcessorConfiguration config, ExecutorService executorService) {
        super(config);
        this.executorService = executorService;
    }
    
    @Override
    public String processItem(String item) throws ProcessingException {
        if (!validateItem(item)) {
            throw new ProcessingException("Invalid item: " + item);
        }
        
        String processed = "processed_" + item.trim().toLowerCase();
        logDebug("Processed item: " + item + " -> " + processed);
        return processed;
    }
    
    /**
     * Process items with custom transformation function.
     * @param items Items to process
     * @param transformer Custom transformation function
     * @return List of transformed results
     */
    public <R> List<R> processWithTransform(Collection<String> items, 
                                          Function<String, R> transformer) {
        return items.stream()
                .filter(this::validateItem)
                .map(item -> {
                    try {
                        String processed = processItem(item);
                        return transformer.apply(processed);
                    } catch (ProcessingException e) {
                        handleError(e, item);
                        return null;
                    }
                })
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }
    
    /**
     * Batch process items with size limits.
     * @param items Items to process
     * @param batchSize Maximum batch size
     * @return List of batch results
     */
    public List<CompletableFuture<List<String>>> processBatches(List<String> items, 
                                                               int batchSize) {
        List<CompletableFuture<List<String>>> futures = new ArrayList<>();
        
        for (int i = 0; i < items.size(); i += batchSize) {
            int endIndex = Math.min(i + batchSize, items.size());
            List<String> batch = items.subList(i, endIndex);
            
            CompletableFuture<List<String>> future = CompletableFuture
                    .supplyAsync(() -> process(batch), executorService);
            futures.add(future);
        }
        
        return futures;
    }
    
    /**
     * Shutdown the processor and cleanup resources.
     */
    public void shutdown() {
        executorService.shutdown();
        try {
            if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                executorService.shutdownNow();
            }
        } catch (InterruptedException e) {
            executorService.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
}

/**
 * Custom exception for processing errors.
 */
public class ProcessingException extends Exception {
    private final String item;
    private final ErrorCode errorCode;
    
    public ProcessingException(String message) {
        super(message);
        this.item = null;
        this.errorCode = ErrorCode.UNKNOWN;
    }
    
    public ProcessingException(String message, String item) {
        super(message);
        this.item = item;
        this.errorCode = ErrorCode.PROCESSING_FAILED;
    }
    
    public ProcessingException(String message, String item, ErrorCode errorCode) {
        super(message);
        this.item = item;
        this.errorCode = errorCode;
    }
    
    public ProcessingException(String message, Throwable cause) {
        super(message, cause);
        this.item = null;
        this.errorCode = ErrorCode.UNKNOWN;
    }
    
    public String getItem() { return item; }
    public ErrorCode getErrorCode() { return errorCode; }
    
    public enum ErrorCode {
        UNKNOWN,
        INVALID_INPUT,
        PROCESSING_FAILED,
        TIMEOUT,
        RESOURCE_UNAVAILABLE
    }
}

/**
 * Utility class with static methods.
 */
public final class ProcessorUtils {
    
    // Private constructor to prevent instantiation
    private ProcessorUtils() {
        throw new UnsupportedOperationException("Utility class");
    }
    
    /**
     * Create processor with default configuration.
     * @return New StringDataProcessor instance
     */
    public static StringDataProcessor createDefaultProcessor() {
        ProcessorConfiguration config = ProcessorConfiguration.builder()
                .name("default")
                .debug(false)
                .timeout(5000)
                .build();
        return new StringDataProcessor(config);
    }
    
    /**
     * Validate collection of items.
     * @param items Items to validate
     * @return true if all items are valid
     */
    public static boolean validateItems(Collection<String> items) {
        return items != null && 
               !items.isEmpty() && 
               items.stream().allMatch(item -> item != null && !item.trim().isEmpty());
    }
    
    /**
     * Generate test data.
     * @param count Number of items to generate
     * @param prefix Prefix for generated items
     * @return List of test items
     */
    public static List<String> generateTestData(int count, String prefix) {
        return IntStream.range(0, count)
                .mapToObj(i -> String.format("%s_%03d", prefix, i))
                .collect(Collectors.toList());
    }
    
    /**
     * Combine multiple processor results.
     * @param results Multiple result lists
     * @return Combined result list
     */
    @SafeVarargs
    public static <T> List<T> combineResults(List<T>... results) {
        return Arrays.stream(results)
                .filter(Objects::nonNull)
                .flatMap(List::stream)
                .collect(Collectors.toList());
    }
}

/**
 * Main class with application entry point.
 */
public class ProcessorApplication {
    
    /**
     * Main method for testing the processor.
     * @param args Command line arguments
     */
    public static void main(String[] args) {
        System.out.println("Starting Processor Application...");
        
        try {
            // Create configuration
            ProcessorConfiguration config = ProcessorConfiguration.builder()
                    .name("main")
                    .debug(true)
                    .timeout(10000)
                    .allowedType("string")
                    .build();
            
            // Create processor
            StringDataProcessor processor = new StringDataProcessor(config);
            
            // Generate test data
            List<String> testData = ProcessorUtils.generateTestData(5, "item");
            
            if (!ProcessorUtils.validateItems(testData)) {
                System.err.println("Invalid test data");
                return;
            }
            
            // Process synchronously
            System.out.println("Processing synchronously...");
            List<String> results = processor.process(testData);
            System.out.println("Results: " + results);
            
            // Process asynchronously
            System.out.println("Processing asynchronously...");
            CompletableFuture<List<String>> futureResults = processor.processAsync(testData);
            List<String> asyncResults = futureResults.get(5, TimeUnit.SECONDS);
            System.out.println("Async results: " + asyncResults);
            
            // Get statistics
            Map<String, Object> stats = processor.getStats();
            System.out.println("Statistics: " + stats);
            
            // Process with transformation
            List<Integer> lengths = processor.processWithTransform(testData, String::length);
            System.out.println("Lengths: " + lengths);
            
            // Cleanup
            processor.shutdown();
            
        } catch (Exception e) {
            System.err.println("Application error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
        
        System.out.println("Application completed successfully.");
    }
}

File: tests/fixtures/sample.c
/**
 * C test fixture for Tree-sitter parsing tests.
 * 
 * This file contains various C language constructs to test comprehensive parsing.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <pthread.h>
#include <unistd.h>

// Preprocessor definitions
#define MAX_ITEMS 100
#define BUFFER_SIZE 1024
#define VERSION "1.0.0"

// Conditional compilation
#ifdef DEBUG
#define LOG(msg) printf("[DEBUG] %s\n", msg)
#else
#define LOG(msg)
#endif

// Type definitions
typedef struct {
    char name[256];
    bool debug;
    int timeout_ms;
    void *options;
} ProcessorConfig;

typedef struct {
    char items[MAX_ITEMS][256];
    int count;
    int capacity;
    int error_count;
    pthread_mutex_t mutex;
} ProcessorResults;

// Function pointer types
typedef int (*ProcessorFunc)(const char *item, char *result, size_t result_size);
typedef void (*ErrorHandler)(const char *error_msg, const char *item);

// Global variables
static ProcessorResults g_results = {0};
static bool g_initialized = false;
static ErrorHandler g_error_handler = NULL;

// Forward declarations
int process_item(const char *item, char *result, size_t result_size);
int process_items(const char **items, int count, char results[][256]);
void *process_async_worker(void *arg);
void handle_error(const char *error_msg, const char *item);
bool validate_item(const char *item);

/**
 * Initialize the processor with configuration
 * @param config Processor configuration
 * @return 0 on success, -1 on error
 */
int processor_init(const ProcessorConfig *config) {
    if (!config) {
        return -1;
    }
    
    if (pthread_mutex_init(&g_results.mutex, NULL) != 0) {
        return -1;
    }
    
    g_results.count = 0;
    g_results.capacity = MAX_ITEMS;
    g_results.error_count = 0;
    g_initialized = true;
    
    LOG("Processor initialized");
    return 0;
}

/**
 * Create default processor configuration
 * @param name Processor name
 * @return Allocated configuration structure (must be freed)
 */
ProcessorConfig* create_default_config(const char *name) {
    ProcessorConfig *config = malloc(sizeof(ProcessorConfig));
    if (!config) {
        return NULL;
    }
    
    strncpy(config->name, name ? name : "default", sizeof(config->name) - 1);
    config->name[sizeof(config->name) - 1] = '\0';
    config->debug = false;
    config->timeout_ms = 5000;
    config->options = NULL;
    
    return config;
}

/**
 * Free processor configuration
 * @param config Configuration to free
 */
void free_config(ProcessorConfig *config) {
    if (config) {
        if (config->options) {
            free(config->options);
        }
        free(config);
    }
}

/**
 * Validate input item
 * @param item Item to validate
 * @return true if valid, false otherwise
 */
bool validate_item(const char *item) {
    if (!item) {
        return false;
    }
    
    size_t len = strlen(item);
    if (len == 0 || len > 255) {
        return false;
    }
    
    // Check for valid characters (alphanumeric and underscore)
    for (size_t i = 0; i < len; i++) {
        char c = item[i];
        if (!((c >= 'a' && c <= 'z') || 
              (c >= 'A' && c <= 'Z') || 
              (c >= '0' && c <= '9') || 
              c == '_' || c == ' ')) {
            return false;
        }
    }
    
    return true;
}

/**
 * Process a single item
 * @param item Input item to process
 * @param result Buffer for result
 * @param result_size Size of result buffer
 * @return 0 on success, -1 on error
 */
int process_item(const char *item, char *result, size_t result_size) {
    if (!item || !result) {
        return -1;
    }
    
    if (!validate_item(item)) {
        if (g_error_handler) {
            g_error_handler("Invalid item", item);
        }
        return -1;
    }
    
    // Process the item (convert to lowercase and add prefix)
    const char *prefix = "processed_";
    size_t prefix_len = strlen(prefix);
    size_t item_len = strlen(item);
    
    if (prefix_len + item_len + 1 > result_size) {
        if (g_error_handler) {
            g_error_handler("Result buffer too small", item);
        }
        return -1;
    }
    
    strcpy(result, prefix);
    
    // Convert to lowercase and append
    for (size_t i = 0; i < item_len; i++) {
        char c = item[i];
        if (c >= 'A' && c <= 'Z') {
            c = c - 'A' + 'a';
        }
        result[prefix_len + i] = c;
    }
    result[prefix_len + item_len] = '\0';
    
    return 0;
}

/**
 * Add result to global results storage
 * @param result Result to add
 * @return 0 on success, -1 on error
 */
int add_result(const char *result) {
    if (!g_initialized) {
        return -1;
    }
    
    pthread_mutex_lock(&g_results.mutex);
    
    if (g_results.count >= g_results.capacity) {
        pthread_mutex_unlock(&g_results.mutex);
        return -1;
    }
    
    strncpy(g_results.items[g_results.count], result, 255);
    g_results.items[g_results.count][255] = '\0';
    g_results.count++;
    
    pthread_mutex_unlock(&g_results.mutex);
    return 0;
}

/**
 * Process multiple items synchronously
 * @param items Array of items to process
 * @param count Number of items
 * @param results Array to store results
 * @return Number of successfully processed items
 */
int process_items(const char **items, int count, char results[][256]) {
    if (!items || !results || count <= 0) {
        return -1;
    }
    
    int processed_count = 0;
    
    for (int i = 0; i < count; i++) {
        char result[256];
        
        if (process_item(items[i], result, sizeof(result)) == 0) {
            strcpy(results[processed_count], result);
            add_result(result);
            processed_count++;
            
            LOG("Item processed successfully");
        } else {
            pthread_mutex_lock(&g_results.mutex);
            g_results.error_count++;
            pthread_mutex_unlock(&g_results.mutex);
        }
    }
    
    return processed_count;
}

/**
 * Structure for passing data to async worker threads
 */
typedef struct {
    const char **items;
    int start_index;
    int end_index;
    char (*results)[256];
    int *processed_count;
    pthread_mutex_t *count_mutex;
} AsyncWorkerArgs;

/**
 * Worker function for asynchronous processing
 * @param arg Pointer to AsyncWorkerArgs structure
 * @return NULL
 */
void *process_async_worker(void *arg) {
    AsyncWorkerArgs *args = (AsyncWorkerArgs *)arg;
    int local_count = 0;
    
    for (int i = args->start_index; i < args->end_index; i++) {
        char result[256];
        
        if (process_item(args->items[i], result, sizeof(result)) == 0) {
            pthread_mutex_lock(args->count_mutex);
            strcpy(args->results[*args->processed_count + local_count], result);
            local_count++;
            pthread_mutex_unlock(args->count_mutex);
            
            add_result(result);
            
            // Simulate processing time
            usleep(1000); // 1ms
        }
    }
    
    pthread_mutex_lock(args->count_mutex);
    *args->processed_count += local_count;
    pthread_mutex_unlock(args->count_mutex);
    
    return NULL;
}

/**
 * Process items asynchronously using multiple threads
 * @param items Array of items to process
 * @param count Number of items
 * @param results Array to store results
 * @param thread_count Number of threads to use
 * @return Number of successfully processed items
 */
int process_items_async(const char **items, int count, char results[][256], int thread_count) {
    if (!items || !results || count <= 0 || thread_count <= 0) {
        return -1;
    }
    
    pthread_t *threads = malloc(thread_count * sizeof(pthread_t));
    AsyncWorkerArgs *args = malloc(thread_count * sizeof(AsyncWorkerArgs));
    pthread_mutex_t count_mutex = PTHREAD_MUTEX_INITIALIZER;
    int processed_count = 0;
    
    if (!threads || !args) {
        free(threads);
        free(args);
        return -1;
    }
    
    int items_per_thread = count / thread_count;
    int remaining_items = count % thread_count;
    
    // Create worker threads
    for (int i = 0; i < thread_count; i++) {
        args[i].items = items;
        args[i].start_index = i * items_per_thread;
        args[i].end_index = args[i].start_index + items_per_thread;
        
        // Distribute remaining items to first few threads
        if (i < remaining_items) {
            args[i].end_index++;
        }
        
        args[i].results = results;
        args[i].processed_count = &processed_count;
        args[i].count_mutex = &count_mutex;
        
        if (pthread_create(&threads[i], NULL, process_async_worker, &args[i]) != 0) {
            // Handle thread creation error
            for (int j = 0; j < i; j++) {
                pthread_join(threads[j], NULL);
            }
            free(threads);
            free(args);
            pthread_mutex_destroy(&count_mutex);
            return -1;
        }
    }
    
    // Wait for all threads to complete
    for (int i = 0; i < thread_count; i++) {
        pthread_join(threads[i], NULL);
    }
    
    free(threads);
    free(args);
    pthread_mutex_destroy(&count_mutex);
    
    return processed_count;
}

/**
 * Get processing statistics
 * @param total_processed Pointer to store total processed count
 * @param error_count Pointer to store error count
 */
void get_stats(int *total_processed, int *error_count) {
    if (!g_initialized) {
        if (total_processed) *total_processed = 0;
        if (error_count) *error_count = 0;
        return;
    }
    
    pthread_mutex_lock(&g_results.mutex);
    if (total_processed) *total_processed = g_results.count;
    if (error_count) *error_count = g_results.error_count;
    pthread_mutex_unlock(&g_results.mutex);
}

/**
 * Reset processor state
 */
void reset_processor(void) {
    if (!g_initialized) {
        return;
    }
    
    pthread_mutex_lock(&g_results.mutex);
    g_results.count = 0;
    g_results.error_count = 0;
    memset(g_results.items, 0, sizeof(g_results.items));
    pthread_mutex_unlock(&g_results.mutex);
}

/**
 * Set error handler callback
 * @param handler Error handler function
 */
void set_error_handler(ErrorHandler handler) {
    g_error_handler = handler;
}

/**
 * Default error handler implementation
 * @param error_msg Error message
 * @param item Item that caused the error
 */
void handle_error(const char *error_msg, const char *item) {
    fprintf(stderr, "Error processing '%s': %s\n", item ? item : "NULL", error_msg);
}

/**
 * Cleanup processor resources
 */
void processor_cleanup(void) {
    if (g_initialized) {
        pthread_mutex_destroy(&g_results.mutex);
        g_initialized = false;
        LOG("Processor cleaned up");
    }
}

/**
 * Generate test data
 * @param items Array to store generated items
 * @param count Number of items to generate
 * @param prefix Prefix for generated items
 */
void generate_test_data(char items[][256], int count, const char *prefix) {
    for (int i = 0; i < count; i++) {
        snprintf(items[i], 256, "%s_%03d", prefix ? prefix : "item", i);
    }
}

/**
 * Main function for testing
 */
int main(int argc, char *argv[]) {
    printf("Starting C processor application...\n");
    
    // Create configuration
    ProcessorConfig *config = create_default_config("main");
    if (!config) {
        fprintf(stderr, "Failed to create configuration\n");
        return 1;
    }
    
    config->debug = true;
    
    // Initialize processor
    if (processor_init(config) != 0) {
        fprintf(stderr, "Failed to initialize processor\n");
        free_config(config);
        return 1;
    }
    
    // Set error handler
    set_error_handler(handle_error);
    
    // Generate test data
    char test_items[5][256];
    generate_test_data(test_items, 5, "item");
    
    // Create array of pointers for processing
    const char *item_ptrs[5];
    for (int i = 0; i < 5; i++) {
        item_ptrs[i] = test_items[i];
    }
    
    // Process synchronously
    printf("Processing synchronously...\n");
    char sync_results[5][256];
    int sync_count = process_items(item_ptrs, 5, sync_results);
    
    printf("Synchronous results (%d items):\n", sync_count);
    for (int i = 0; i < sync_count; i++) {
        printf("  %s\n", sync_results[i]);
    }
    
    // Reset for async processing
    reset_processor();
    
    // Process asynchronously
    printf("Processing asynchronously...\n");
    char async_results[5][256];
    int async_count = process_items_async(item_ptrs, 5, async_results, 2);
    
    printf("Asynchronous results (%d items):\n", async_count);
    for (int i = 0; i < async_count; i++) {
        printf("  %s\n", async_results[i]);
    }
    
    // Get statistics
    int total_processed, error_count;
    get_stats(&total_processed, &error_count);
    printf("Statistics: Processed=%d, Errors=%d\n", total_processed, error_count);
    
    // Cleanup
    processor_cleanup();
    free_config(config);
    
    printf("C processor application completed successfully.\n");
    return 0;
}